### ISSUE TITLE
Associativity of vector concatenation REWRITE sometimes doesn't apply

### ISSUE THREAD

**Issue by NathanielB123** [agda/agda#7602]
The `REWRITE` rule for associativity for length-indexed `Vec`tor concatenation doesn't seem to apply in not-fully-general cases.

```agda
{-# OPTIONS --rewriting #-}

open import Agda.Builtin.Equality using (_≡_; refl)
import Agda.Builtin.Equality.Rewrite

module RewriteFailMinimal where

data ℕ : Set where
  ze : ℕ
  su : ℕ → ℕ

variable
  x y z : ℕ
  A : Set

_+_ : ℕ → ℕ → ℕ
ze     + y = y
(su x) + y = su (x + y)

data Vec (A : Set) : ℕ → Set where
  []  : Vec A ze
  _∷_ : A → Vec A x → Vec A (su x)

variable
  xs ys zs : Vec A x

_++_ : Vec A x → Vec A y → Vec A (x + y)
[]       ++ ys = ys
(x ∷ xs) ++ ys = x ∷ (xs ++ ys)

postulate
  +-assoc : (x + y) + z ≡ x + (y + z)

{-# REWRITE +-assoc #-}

postulate
  ++-assoc : (xs ++ ys) ++ zs ≡ xs ++ (ys ++ zs)

{-# REWRITE ++-assoc #-}

ohdear : ∀ {xs : Vec A ze} → (xs ++ ys) ++ zs ≡ xs ++ (ys ++ zs)
ohdear = {!!}
```

`--local-confluence-check` is not happy with `++-assoc` I assume because of these sorts of situations:
```
Error
/home/nathaniel/agda/fyp/RewriteFailMinimal.agda:39,13-21
Local confluence check failed: (xs ++ ys) ++ zs reduces to both
xs ++ (ys ++ zs) and (xs ++ ys) ++ zs which are not equal because
x != x + ys.x of type ℕ
when checking confluence of the rewrite rule ++-assoc with
RewriteFailMinimal._+_-clause2
Error
/home/nathaniel/agda/fyp/RewriteFailMinimal.agda:39,13-21
Local confluence check failed: (xs ++ ys) ++ zs reduces to both
xs ++ (ys ++ zs) and (xs ++ ys) ++ zs which are not equal because
ze != ys.x of type ℕ
when checking confluence of the rewrite rule ++-assoc with
RewriteFailMinimal._+_-clause1
Error
/home/nathaniel/agda/fyp/RewriteFailMinimal.agda:39,13-21
Local confluence check failed: (xs ++ ys) ++ zs reduces to both
xs ++ (ys ++ zs) and (xs ++ ys) ++ zs which are not equal because
y != y + ys.x of type ℕ
when checking confluence of the rewrite rule ++-assoc with +-assoc
```

~~I assume the underlying problem has something to do with the desired REWRITE depending on an earlier REWRITE to typecheck, and the type of the to-be-rewritten term being more specific than that of the one in original `REWRITE` (i.e. the type of `(xs ++ ys) ++ zs` in `ohdear` is `Vec A ((ze + y) + z) = Vec A (y + z)` which is not obviously of the form `Vec A ((x + y) + z)`, but interestingly, a seemingly quite similar scenario with `++[] : xs ++ [] ≡ xs` appears to work out just fine:~~
EDIT: As explained by Jesper below, the issue here is the implicit `x + y` length argument. Agda rewrites are not type directed, so the type is not really relevant.

```agda
postulate
  +0 : x + ze ≡ x

{-# REWRITE +0 #-}

postulate
  ++[] : xs ++ [] ≡ xs

{-# REWRITE ++[] #-}

works : ∀ {xs : Vec A ze} → xs ++ [] ≡ xs
works = refl
```

Is this a known limitation or bug?

(Tested on Agda version 2.7.0.1)

**Comment by jespercockx** [comment]
Well, the problem here really is that the (implicit) length arguments to `_++_` become part of the pattern of the `++-assoc` rewrite rule, i.e. the pattern becomes `_++_ {A} {m + n} {o} (_++_ {A} {m} {n} xs ys) zs`. Hence the rewrite rule will *only* fire if the argument is syntactically `m + n` where `m` is the length of `xs` and `n` the length of `ys`. But this then creates critical pairs with the clauses and rewrite rules for `_+_`.

You could try to resolve those by adding additional rewrite rules, e.g. `++-ze : _++_ {A} {ze} {n} xs ys ≡ ys`. But resolving the overlap with the `su` case and with associativity for `_+_` is more problematic. In particular, adding them just leads down a rabbit hole of increasingly complex critical pairs:
```agda
Local confluence check failed:
_++_ {A} {su ((m + n) + n₁)} {o} (_++_ {A} {su (m + n)} {n₁} xs ys)
zs
reduces to both
_++_ {A} {su (m + n)} {n₁ + o} xs (_++_ {A} {n₁} {o} ys zs) and
_++_ {A} {su (m + (n + n₁))} {o} (_++_ {A} {su (m + n)} {n₁} xs ys)
```

Ideally there would be some way to make the length arguments irrelevant for Agda so it only matches the vector arguments. In principle that is what shape-irrelevance (which is hidden under `--experimental-irrelevance`) is for but a quick test seems to indicate that it does not play well together with rewriting.

So in short, there doesn't seem to be anything obviously wrong in what Agda is doing, but this might just be a fundamental limitation of combining dependent types like Vec with rewrite rules.

**Comment by NathanielB123** [comment]
Ahh ok, this seems a bit unfortunate, but makes sense, thanks (and I guess explains why `∀ {xs : Vec A ze} → xs ++ [] ≡ xs` worked - the corresponding rewrite doesn't expect a neutral implicit arg like `x + y`; I guess matching `x` with `ze` is much easier than `x + y` with `y`). 

To be fair also, in my original use case, the datatype is actually indexed by a boolean rather than a natural number and therefore I think the set of additional specialised rewrites I need is finite, so this isn't the end of the world.

**Comment by andreasabel** [comment]
The question seems answered, closing.
